### DISCUSSION TITLE
"appdirs" missing

### DISCUSSION THREAD

**Discussion by forrie** [bernd-wechner/Degoo#35] [General]
I am just starting to test this out.   I installed from requirements.txt -- MacOS Monterey under Macports.   pip and pythong are 3.9:

```
degoo git:(master) ✗ ./degoo_ls
Traceback (most recent call last):
  File "./degoo_ls", line 21, in <module>
    import degoo
  File "/usr/local/src/degoo/degoo/__init__.py", line 22, in <module>
    from appdirs import user_config_dir
ModuleNotFoundError: No module named 'appdirs'

root@MacbookPro ➜  degoo git:(master) ✗ pip install appdirs

Requirement already satisfied: appdirs in /opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages (1.4.4)
```
This is just output from degoo_ls, but degoo_login does the same thing, it's here for context.

What am I missing?

Thanks!

**Comment by bernd-wechner** [comment]
That is odd indeed.

What I can recommend is trying something like this:

```
bernd@bigfoot:~$ python3
Python 3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import appdirs
>>> appdirs.__version__
'1.4.4'
``` 
If that fails, and appdirs is not found try this:

```
>>> import sys
>>> sys.path
['', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/home/bernd/.local/lib/python3.8/site-packages', '/usr/local/lib/python3.8/dist-packages', '/usr/lib/python3/dist-packages']
```

You're using a Mac by looks of it and your prompts are tad interesting ;-), but the install dir is a little unfamiliar to me too.

Essentially if `/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages (1.4.4)` isn't in `sys.path` then the import will fail. 

Run contexts can vary wildly alas, so one way to real sure of the the `sys.path` that the degoo script see is to print it out before trying the appdirs import. 

Hope that helps. It's always hard to be sure without poking around more, and looks like a Python environment issue rather than anything to do with this scrip per se. 

Note that pip can accept and install dir too.

https://pip.pypa.io/en/latest/cli/pip_install/#cmdoption-t

**Comment by forrie** [comment]
Thanks for responding.   Here's what I found with the above.

`root@MacbookPro ➜ degoo git:(master) ✗ python3

Python 3.9.7 (default, Oct 30 2021, 16:26:10)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import appdirs

[ that worked ]


>>> import sys
>>> sys.path
['', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python39.zip', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/readline', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload', '/Users/faldrich/Library/Python/3.9/lib/python/site-packages', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages', '/usr/local/src/nikola', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/gitup-0.5.1-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/GitPython-3.1.15-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/gitdb-4.0.7-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/smmap-4.0.0-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/ansible_core-2.12.0.dev0-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/colorama-0.4.3-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/s3transfer-0.5.0-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/docutils-0.15.2-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/jmespath-0.10.0-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/awscli-1.21.9-py3.9.egg', '/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/botocore-1.22.9-py3.9.egg']`

I'm somewhat familiar with Python, so I may be missing something here.  I'm using Macports for Python3:

`python --version
Python 3.9.7`

**Comment by forrie** [comment]
I think I figured out how to get around it.  Your code is presently calling a hard-coded path /usr/bin/python3, where Macports is in /opt/local/bin/python3.   So, even if you modify the top of one file (ie: degoo_ls), the other code is still calling a different location.

So to get around this:

```
➜ degoo git:(master) ✗ python3 degoo_login

username: [ my email addy redacted ]

Password:

Login failed with: 429: Too Many Requests

Login failed.
```

This is weird, and I double-checked my password and typed it in very carefully.

Another note here, I think really the same issue applies even when calling directly (initially) with python3, the other code is referencing another location and associated libs.   In this case, I got stuck in a loop with degoo_login and it would not let me CTRL-D or CTRL-C out of it, I had to kill the shell :)

**Comment by bernd-wechner** [comment]
>  Your code is presently calling a hard-coded path

Not calling it per se. What you're referring to there is called a shebang.  It's just a convention to let a shell know what program to run, but no shell is obliged to  respect it, and it's not a Python thing. See more here:

See here: 

https://en.wikipedia.org/wiki/Shebang_(Unix)

That said a more portable shebang I think is probably:

```
#!/usr/bin/env python3
```

But to be honest I've not given it much thought as I don't pot much, and when I do it's Linux/Windows, I don't have any Mac experience.

If that works better as shebang for you on a Mac, I can try it her eon Linux to and use that shebang in the hope it's more portable.

**Comment by forrie** [comment]
I think coding this in a portable manner is going to be better for a larger audience.  So I would vote for the shebang bang bang :)

Thanks again!

**Comment by yopines** [comment]
I too purchased the 10TB lifetime package to suddenly realize that it cannot be used for General Backup.

Having a tool for MAC, Windows,  Mobile would be great - let's start with a **MAC** as it is the closest you get Linux.

The initial tool can be very simple and then expanded.

1.  Create a password protected Backup Container on the Degoo Server - this is your backup space for Device # 1, 2,...  n
2. Option 1 - full image backup - just backup everything on your hard-drive - bits and bytes
3. Option 2 - Backup & Sync - choose on your drive which directories to backup and keep them in sync with your degoo container.


That's it for now.  I am willing to send a little cash donation to get this thing completed before end of 2021.

Let me know
Thanks
